### PR TITLE
improvement: more general support for half-closable connections

### PR DESCRIPTION
**Pull request by onokonem**

### CODE REVIEW

File: https.go [hunk 1]
```diff
@@ -64,6 +64,14 @@ func (proxy *ProxyHttpServer) connectDial(network, addr string) (c net.Conn, err
 	return proxy.ConnectDial(network, addr)
 }
 
+type halfClosable interface {
+	net.Conn
+	CloseWrite() error
+	CloseRead() error
+}
+
+var _ halfClosable = (*net.TCPConn)(nil)
```
Comment (elazarl) [inline https.go]: Why is that needed?

Comment (onokonem) [reply https.go]: this one is just to check the interface is complaint to the needs.

**Comment by onokonem** [comment]
> Looks nice, but I wonder, what is the usecase?

actually, 2 cases (and I run into both in our project)

1. Not only net.TCPConn is half-closable, but also the net.UnixConn
2. sometimes we have to wrap net.TCPConn with some custom type (we are reporting traffic, for example)
